src/rest/parser.hpp

```diff
@@ -0,0 +1,59 @@
+/*
+ *  Copyright (c) 2020, The OpenThread Authors.
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions are met:
+ *  1. Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *  2. Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *  3. Neither the name of the copyright holder nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   This file includes parser definition for RESTful HTTP server.
+ */
+
+#ifndef OTBR_REST_PARSER_HPP_
+#define OTBR_REST_PARSER_HPP_
+
+#include <memory>
+
+#include "rest/request.hpp"
+
+namespace otbr {
+namespace rest {
+
+class Parser
+{
+public:
+    Parser(Request *aRequest);
+    void Init();
+    void Process(const char *aBuf, int aLength);
+
+private:
+    std::shared_ptr<void> mParser;
```

Comment:
why not use the mParser object itself? For example `http_parser_t mParser;`?
You mentioned that there is header inclusion problems when this `parser.hpp` is included by other files. But I don't understand why this `parser.hpp` file will be included by other files other than `connection.hpp`.

Reply:
No, it because when building in some other environment, It said `http_parser.h` can't be found, but actually I didn't use it in other place and declare as a public library when using it. then I consider use it as how we use cjson

Reply:
you include `parser.hpp` in `connection.hpp`, and include `connection.hpp` in `rest_web_server.hpp`. you can link `http_parser` as a public library of target `otbr_rest`  to solve it:
```c++
target_link_libraries(otbr_rest PUBLIC http_parser)
```